Documentation for raft ANN benchmark containers. (#1833)
rapidsai/raft: docs/source/raft_ann_benchmarks.md

This project provides a benchmark program for various ANN search implementations. It's especially suitable for comparing GPU implementations as well as comparing GPU against CPU.

## Table of Contents

- [Installing and Running the Benchmarks](#installing--and-running-the-benchmarks)
    - [Using conda](#conda)

>       - [End-to-end example: Million-scale](end-to-end-example-million-scale)

Reviewer: It seems a little akward that we have the million-scale example embedded underneath "using conda" while we have the billion-scale as its own top-level. Should we move the million-scale to be just before the billion-scale?

### Docker

We provide images for GPU enabled systems, as well as systems without a GPU. The following images are available:

- `raft-ann-bench`: Contains GPU and CPU benchmarks, can run all algorithms supported. Will download million-scale datasets as required. Best suited for users that prefer a smaller container size for GPU based systems. Requires the NVIDIA Container Toolkit to run GPU algorithms, can run CPU algorithms without it.
- `raft-ann-bench-datasets`: Contains the GPU and CPU benchmarks with million-scale datasets already included in the container. Best suited for users that want to run multiple million scale datasets already included in the image.
- `raft-ann-bench-cpu`: Contains only CPU benchmarks with minimal size. Best suited for users that want the smallest containers to reproduce benchmarks on systems without a GPU.

Nightly images are located in [dockerhub](https://hub.docker.com/r/rapidsai/raft-ann-bench), meanwhile release (stable) versions are located in [NGC](https://hub.docker.com/r/rapidsai/raft-ann-bench), starting with release 23.10.

> - To pull the nightly containers the format is the following:

Reviewer: ```suggestion
- The following command pulls the nightly container for python version 10, cuda version 12, and RAFT version 23.10:
```

```bash
docker pull rapidsai/raft-ann-bench:23.10a-cuda12.0-py3.10 #substitute raft-ann-bench for the exact desired container.
```

The CUDA and python versions can be changed for the supported values:

Supported CUDA versions: 11.8 and 12.0
Supported Python versions: 3.9 and 3.10.

> **Note:** GPU containers use the CUDA toolkit from inside the container, the only requirement is a driver that supports that version. So, for example, CUDA 11.8 containers can run in systems with a CUDA 12.x capable driver.

Reviewer: ```suggestion
**Note:** GPU containers use the CUDA toolkit from inside the container, the only requirement is a driver installed on the host machine that supports that version. So, for example, CUDA 11.8 containers can run in systems with a CUDA 12.x capable driver.
```

> - To pull the release containers the format is the following (only available after RAPIDS 23.10 release):

Reviewer: ```suggestion
-  The following command (only available after RAPIDS 23.10 release) pulls the container:
```

23.08 doesn't work here right? Should this be 23.10?

```bash
docker pull nvcr.io/nvidia/rapidsai/raft-ann-bench:23.08-cuda11.8-py3.10 #substitute raft-ann-bench for the exact desired container.
```

#### Container Usage

> The containers can be used in two manners:

Reviewer: ```suggestion
The container can be used in two different ways:
```

> 1. **Quick benchmark with single `docker run`**: The docker containers already include helper scripts to be able to invoke most of the functionality of the benchmarks from docker run for a simple and easy way to run benchmarks. To use the containers in this manner, use the following commands:

Reviewer: ```suggestion
1. **Automated benchmark with single `docker run` (ease mode)**: Helper scripts are included to ease the procedure of running benchmarks end-to-end:
```

> For GPU systems, where `$DATA_FOLDER` is a local folder where you want datasets stored in $DATA_FOLDER/datasets and results in `$DATA_FOLDER/result` (we highly recommend `$DATA_FOLDER` to be a dedicated folder for the datasets and results of the containers):

Reviewer: ```suggestion
For GPU systems, where `$DATA_FOLDER` is a local folder where you want datasets stored in `$DATA_FOLDER/datasets` and results in `$DATA_FOLDER/result` (we highly recommend `$DATA_FOLDER` to be a dedicated folder for the datasets and results of the containers):
```

```bash

> docker run --gpus all --rm -it \

Reviewer: Can you include the export of DATA_FOLDER in this command too?

    -v $DATA_FOLDER:/home/rapids/benchmarks  \
    rapidsai/raft-ann-bench:23.10a-cuda11.8-py3.10 \

>     "--dataset deep-image-96-angular" \

Reviewer: Is it possible for them to specify their own directory of configuration files?

    "--normalize" \
    "--algorithms raft_cagra,raft_ivf_pq" \
    ""
```

Where:

```bash

> docker run --gpus all --rm -it \

Reviewer: Could we maybe break this into a table or maybe print usage and include the comments for the docker-run-specific parts in the command above? Just a thought. It doesn't look horrible but I think it's dense w/ information.

    -v $DATA_FOLDER:/home/rapids/benchmarks  \ # <- local folder to store datasets and results
    rapidsai/raft-ann-bench:23.10a-cuda11.8-py3.10 \ # <- image to use, either `raft-ann-bench` or `raft-ann-bench-datasets`, can choose RAPIDS, cuda and python versions.
    "--dataset deep-image-96-angular" \ # <- dataset name
    "--normalize" \ # <- whether to normalize the dataset, leave string empty ("") to not normalize.
    "--algorithms raft_cagra" \ # <- what algorithm(s) to use as a ; separated list, as well as any other argument to pass to `raft_ann_benchmarks.run`
    "" # optional arguments to pass to `raft_ann_benchmarks.plot`
```

For CPU systems the same interface applies, except for not needing the gpus argument and using the cpu images:
```bash
docker run  all --rm -it \
    -v $DATA_FOLDER:/home/rapids/benchmarks  \
    rapidsai/raft-ann-bench-cpu:23.10a-py3.10 \
     "--dataset deep-image-96-angular" \
     "--normalize" \
     "--algorithms raft_cagra" \
     ""
```

> **Note:** The user inside the containers is `root`, to workaround this the scripts in the containers fix the user of the output files after the benchmarks are run. If the benchmarks are interrupted, the owner of the datasets/results produced by the container will be wrong, and can be fixed by the user.

Reviewer: ```suggestion
**Note:** The user inside the containers is `root`. To workaround this, the scripts in the containers fix the user of the output files after the benchmarks are run. If the benchmarks are interrupted, the owner of the `datasets/results` produced by the container will be wrong, and will need to be manually fixed by the user.
```

>  and can be fixed by the user.
Does this mean they will need to be fixed manually by the user?

> 2. **Using the preinstalled `raft_ann_benchmarks` python package**: The docker containers are built using the conda packages described in the following section, so they can be used directly as if they were installed manually following the instructions in the next section. This is the option that allows the full flexibility of the benchmarking scripts, and recommended for advanced users. This allows using the full flexibility of the scripts. To use the python scripts directly, an easy way is to use the following command:

Reviewer: Would it make sense to explain the conda install first and then the docker install? That way instead of saying "in the next section" where users haven't yet been exposed to the content, we can say "as outlined in the previous section" .

Author: Indeed, changed the order of the sections

> 2. **Using the preinstalled `raft_ann_benchmarks` python package**: The docker containers are built using the conda packages described in the following section, so they can be used directly as if they were installed manually following the instructions in the next section. This is the option that allows the full flexibility of the benchmarking scripts, and recommended for advanced users. This allows using the full flexibility of the scripts. To use the python scripts directly, an easy way is to use the following command:

Reviewer: ```suggestion
2. **Using the preinstalled `raft_ann_benchmarks` python package (advanced mode)**: The docker containers are built using the conda packages described in the following section, so they can be used directly as if they were installed manually following the instructions in the next section. This is recommended for advanced users, and is the option that allows the full flexibility of the benchmarking scripts. To use the python scripts directly, use the following command:
```

```bash
docker run --gpus all --rm -it \
    -v $DATA_FOLDER:/home/rapids/benchmarks  \
    rapidsai/raft-ann-bench:23.10a-cuda11.8-py3.10 \
    --entrypoint /bin/bash
```

This will drop you into a command line in the container, with the `raft_ann_benchmarks` python package ready to use:

```
(base) root@00b068fbb862:/home/rapids#
```

> Additionally, the containers could be run in dettached form without any issue.

Reviewer: ```suggestion
Additionally, the containers could be run in dettached mode without any issue.
```

> For details on how to use the python package from the command line, see the [corresponding section](#python-package-usage).

Reviewer: ```suggestion
See the [python package usage](#python-package-usage) for more details on how to use the python package.
```

- `fashion-mnist-784-euclidean`
- `glove-50-angular`
- `glove-100-angular`

> - `lastfm-65-angular`

Reviewer: Is this one not working?

Author: nope, at least not downloaded from the original location as we had it. We can add it back in the scripts and then add it back here as a follow up
